src/roswire/common/source.py

```diff
@@ -134,6 +107,170 @@ def for_app_instance(
+                    executables,
+                    package)
+            if cmd == "add_subdirectory":
+                executables = self.__process_add_subdirectory(
+                    args,
+                    cmake_env,
+                    executables,
+                    package)
+        return CMakeInfo(executables)
+
+    def __process_add_subdirectory(
+        self,
+        args: t.List[str],
+        cmake_env: t.Dict[str, str],
+        executables: t.Dict[str, ExecutableInfo],
+        package: Package
+    ) -> t.Dict[str, ExecutableInfo]:
+        new_env = cmake_env.copy()
+        new_env['cwd'] = os.path.join(cmake_env.get('cwd', '.'), args[0])
+        join = os.path.join(package.path, new_env['cwd'])
+        cmakelists_path = os.path.join(join, 'CMakeLists.txt')
+        logger.debug(f"Processing {cmakelists_path}")
+        included_package_info = self.process_cmake_contents(
+            self._files.read(cmakelists_path),
+            package,
+            new_env,
+        )
+        executables = {
+            **executables,
+            **{s: included_package_info.executables[s]
+               for s in included_package_info.executables}
+        }
+        return executables
+
+    def __process_add_executable(
+        self,
+        args: t.List[str],
+        cmake_env: t.Dict[str, str],
+        executables: t.Dict[str, ExecutableInfo],
+        package: Package
+    ) -> None:
+        name = args[0]
+        sources: t.Set[str] = set()
+        for source in args[1:]:
+            if 'cwd' in cmake_env:
+                sources.add(os.path.join(cmake_env['cwd'], source))
+            else:
+                sources.add(source)
+        logger.debug(f"Adding C++ sources for {name}")
+        executables[name] = ExecutableInfo(
+            name=name,
+            language=SourceLanguage.CXX,
+            kind=ExecutableKind.NODE,
+            sources=sources,
+            restrict_to_paths=self.package_paths(package))
+
+    def __process_add_library(
+        self,
+        args: t.List[str],
+        cmake_env: t.Dict[str, str],
+        executables: t.Dict[str, ExecutableInfo],
+        package: Package
+    ) -> None:
+        name = args[0]
+        if 'cwd' in cmake_env:
+            sources = {os.path.join(cmake_env['cwd'], s) for s in args[1:]}
+        else:
+            sources = set(args[1:])
+        logger.debug(f"Adding C++ library {name}")
+        executables[name] = ExecutableInfo(
+            name,
+            SourceLanguage.CXX,
+            ExecutableKind.LIBRARY,
+            sources,
+            self.package_paths(package))
+
+    def __process_python_executables(
+        self,
+        args: t.List[str],
+        cmake_env: t.Dict[str, str],
+        executables: t.Dict[str, ExecutableInfo]
+    ) -> None:
+        opts, args = cmake_argparse(
+            args,
+            {"PROGRAMS": "*", "DESTINATION": "*"}
+        )
+        if 'PROGRAMS' in opts:
```

Comment:
prefer `not in` and then the code below can be unindented for better readability

Reply:
Done